### PR TITLE
Precisely format large numbers using Intl through i18n

### PR DESCRIPTION
**Pull request by lukaw3d**
Alternative to / opposite of https://github.com/oasisprotocol/explorer/pull/333

Consistently format even large stringified numbers using native Intl

TODO: replace all toLocaleString and `<span>{param.value as string}</span>`

### CODE REVIEW

File: src/locales/__tests__/getPreciseNumberFormat.test.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  const options = getPreciseNumberFormat(props.value)
+  return t('common.valuePair', options)
+}
+
+describe('getPreciseNumberFormat', () => {
+  test('should render precise number with formatting', () => {
+    const { result } = renderHook(usePreciseNumberFormat, {
+      initialProps: { value: '111222333444555666777888999.111222333444555666' },
+    })
+    expect(result.current).toBe('111,222,333,444,555,666,777,888,999.111222333444555666')
```
Comment (lukaw3d) [inline src/locales/__tests__/getPreciseNumberFormat.test.ts:16]: Big problem with this approach: Firefox does not support stringified number in Intl.NumberFormat accurately and returns "111,222,333,444,555,670,000,000,000" (like NodeJS 18)

Comment (lukaw3d) [reply src/locales/__tests__/getPreciseNumberFormat.test.ts:16]: Same issue on Safari

Comment (lukaw3d) [reply src/locales/__tests__/getPreciseNumberFormat.test.ts:16]: I think Firefox fix will come with https://bugzilla.mozilla.org/show_bug.cgi?id=1795756

Comment (lukaw3d) [reply src/locales/__tests__/getPreciseNumberFormat.test.ts:16]: Safari seems to have missed it in https://bugs.webkit.org/show_bug.cgi?id=215438

Comment (lukaw3d) [reply src/locales/__tests__/getPreciseNumberFormat.test.ts:16]: Made it fallback to "111222333444555666777888999.111222333444555666" in Firefox and Safari. E2E tests verify it

File: playwright/tests/getPreciseNumberFormat.spec.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+  test('exceeding maximumFractionDigits:20 should fallback to precise unformatted number', async ({
+    page,
+  }) => {
+    await setup(page, '111222333444555666777888999111222333444555666', 36)
+    await expect(
+      page.getByText('111222333.444555666777888999111222333444555666', { exact: true }),
+    ).toBeVisible()
+  })
+})
```
Comment (lukaw3d) [inline playwright/tests/getPreciseNumberFormat.spec.ts:80]: How it behaves across browsers

@donouwens says "Yeah, I think this should work. :+1:"

File: src/locales/en/translation.json [hunk 3]
```diff
@@ -95,9 +95,10 @@
     "undefined": "Undefined",
     "unknown": "Unknown",
     "value": "Value",
-    "valueInToken": "{{value}} {{ticker}}",
-    "roundedValueInToken": "{{value}}… <TickerLink />",
-    "valueInTokenWithLink": "{{value}} <TickerLink />",
+    "valueLong": "{{value, number}}",
+    "valueInToken": "{{value, number}} {{ticker}}",
+    "roundedValueInToken": "{{value, number}}… <TickerLink />",
+    "valueInTokenWithLink": "{{value, number}} <TickerLink />",
```
Comment (lukaw3d) [inline src/locales/en/translation.json:101]: I don't know how to better name these and lessThanAmount and valuePair. Future issue?

File: src/locales/getPreciseNumberFormat.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+const isPreciseIntlSupported =
+  new Intl.NumberFormat('en-US', { maximumFractionDigits: 20 })
+    // @ts-expect-error Missing types for NumberFormat v3
+    .format('111222333444555666777888999.111222333444555666')
+    // Chrome 115 and Firefox 116 return '111,222,333,444,555,666,777,888,999.111222333444555666'
+    // Firefox 115 and Safari 16.5 return '111,222,333,444,555,670,000,000,000'
+    .at(-1) === '6'
+
+export function getPreciseNumberFormat(value: string) {
```
Comment (lukaw3d) [inline src/locales/getPreciseNumberFormat.ts:11]: Also open for name suggestions

File: src/locales/en/translation.json [hunk 5]
```diff
@@ -95,9 +95,10 @@
     "undefined": "Undefined",
     "unknown": "Unknown",
     "value": "Value",
-    "valueInToken": "{{value}} {{ticker}}",
-    "roundedValueInToken": "{{value}}… <TickerLink />",
-    "valueInTokenWithLink": "{{value}} <TickerLink />",
+    "valueLong": "{{value, number}}",
```
Comment (lukaw3d) [inline src/locales/en/translation.json:98]: Here's what happens if we add a language switcher:
| English | French |
| --- | --- |
| localhost_1234_mainnet_sapphire_address_0xE8D562606F35CB14dA3E8faB1174F9B5AE8319c4_tokens_erc-20 (1)  | localhost_1234_mainnet_sapphire_address_0xE8D562606F35CB14dA3E8faB1174F9B5AE8319c4_tokens_erc-20  |

**Review by csillag — APPROVED** [review]
This whole situation sucks. I think  this probably the best we can do in these circumstances. Handles all corner cases nicely. LGTM.

**Pull request merged**